File: internal/services/s3/store_multipart.go
package s3

import (
	"bytes"
	"context"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"time"
)

func (s *FileBucketStore) CreateMultipartUpload(ctx context.Context, input CreateMultipartUploadInput) (MultipartUpload, error) {
	if err := ctx.Err(); err != nil {
		return MultipartUpload{}, err
	}
	if err := s.requireBucketAndKey(ctx, input.Bucket, input.Key); err != nil {
		return MultipartUpload{}, err
	}
	upload := MultipartUpload{
		Bucket:             input.Bucket,
		Key:                input.Key,
		UploadID:           newUploadID(),
		CreatedAt:          time.Now().UTC(),
		ContentType:        input.ContentType,
		ContentEncoding:    input.ContentEncoding,
		CacheControl:       input.CacheControl,
		ContentDisposition: input.ContentDisposition,
		Metadata:           cleanMetadata(input.Metadata),
		Encryption:         cleanServerSideEncryption(input.Encryption),
	}
	if upload.ContentType == "" {
		upload.ContentType = "application/octet-stream"
	}
	path := s.multipartUploadPath(input.Bucket, upload.UploadID)
	if err := os.MkdirAll(filepath.Join(path, "parts"), 0o755); err != nil {
		return MultipartUpload{}, fmt.Errorf("create multipart upload: %w", err)
	}
	if err := writeJSONFile(filepath.Join(path, "upload.json"), upload); err != nil {
		return MultipartUpload{}, err
	}
	return upload, nil
}

func (s *FileBucketStore) UploadPart(ctx context.Context, bucket string, key string, uploadID string, partNumber int, body io.Reader, contentMD5 string) (MultipartPart, error) {
	if err := ctx.Err(); err != nil {
		return MultipartPart{}, err
	}
	if partNumber < 1 || partNumber > 10000 {
		return MultipartPart{}, fmt.Errorf("invalid part number")
	}
	upload, ok, err := s.getMultipartUpload(ctx, bucket, key, uploadID)
	if err != nil {
		return MultipartPart{}, err
	}
	if !ok || upload.Key != key {
		return MultipartPart{}, fmt.Errorf("multipart upload does not exist")
	}
	data, err := io.ReadAll(body)
	if err != nil {
		return MultipartPart{}, fmt.Errorf("read multipart part: %w", err)
	}
	if err := validateContentMD5(contentMD5, data); err != nil {
		return MultipartPart{}, err
	}
	sum := md5.Sum(data)
	part := MultipartPart{
		PartNumber:   partNumber,
		ETag:         `"` + hex.EncodeToString(sum[:]) + `"`,
		Size:         int64(len(data)),
		LastModified: time.Now().UTC(),
	}
	path := s.multipartPartPath(bucket, uploadID, partNumber)
	if err := os.MkdirAll(path, 0o755); err != nil {
		return MultipartPart{}, fmt.Errorf("create multipart part: %w", err)
	}
	if err := os.WriteFile(filepath.Join(path, "body"), data, 0o644); err != nil {
		return MultipartPart{}, fmt.Errorf("write multipart part body: %w", err)
	}
	if err := writeJSONFile(filepath.Join(path, "part.json"), part); err != nil {
		return MultipartPart{}, err
	}
	return part, nil
}

func (s *FileBucketStore) ListParts(ctx context.Context, bucket string, key string, uploadID string) (MultipartUpload, []MultipartPart, bool, error) {
	if err := ctx.Err(); err != nil {
		return MultipartUpload{}, nil, false, err
	}
	upload, ok, err := s.getMultipartUpload(ctx, bucket, key, uploadID)
	if err != nil || !ok {
		return upload, nil, ok, err
	}
	parts, err := s.readMultipartParts(bucket, uploadID)
	if err != nil {
		return MultipartUpload{}, nil, false, err
	}
	return upload, parts, true, nil
}

func (s *FileBucketStore) CompleteMultipartUpload(ctx context.Context, bucket string, key string, uploadID string, partNumbers []int) (Object, bool, error) {
	if err := ctx.Err(); err != nil {
		return Object{}, false, err
	}
	upload, ok, err := s.getMultipartUpload(ctx, bucket, key, uploadID)
	if err != nil || !ok {
		return Object{}, ok, err
	}
	var combined bytes.Buffer
	partETags := make([]string, 0, len(partNumbers))
	for _, partNumber := range partNumbers {
		var part MultipartPart
		if err := readJSONFile(filepath.Join(s.multipartPartPath(bucket, uploadID, partNumber), "part.json"), &part); err != nil {
			if os.IsNotExist(err) {
				return Object{}, true, fmt.Errorf("multipart part %d does not exist", partNumber)
			}
			return Object{}, true, fmt.Errorf("read multipart part metadata: %w", err)
		}
		body, err := os.ReadFile(filepath.Join(s.multipartPartPath(bucket, uploadID, partNumber), "body"))
		if err != nil {
			if os.IsNotExist(err) {
				return Object{}, true, fmt.Errorf("multipart part %d does not exist", partNumber)
			}
			return Object{}, true, fmt.Errorf("read multipart part: %w", err)
		}
		combined.Write(body)
		partETags = append(partETags, part.ETag)
	}
	object, err := s.PutObject(ctx, PutObjectInput{
		Bucket:             upload.Bucket,
		Key:                upload.Key,
		Body:               bytes.NewReader(combined.Bytes()),
		ContentType:        upload.ContentType,
		ContentEncoding:    upload.ContentEncoding,
		CacheControl:       upload.CacheControl,
		ContentDisposition: upload.ContentDisposition,
		Metadata:           upload.Metadata,
		Encryption:         upload.Encryption,
	})
	if err != nil {
		return Object{}, true, err
	}
	object.ETag = multipartETag(partETags)
	if err := writeObjectMetadata(filepath.Join(s.objectPath(upload.Bucket, upload.Key), "object.json"), object); err != nil {
		return Object{}, true, err
	}
	if object.VersionID != "" {
		if err := s.writeObjectVersion(s.objectPath(upload.Bucket, upload.Key), object, combined.Bytes()); err != nil {
			return Object{}, true, err
		}
	}
	if err := os.RemoveAll(s.multipartUploadPath(bucket, uploadID)); err != nil {
		return Object{}, true, fmt.Errorf("delete completed multipart upload: %w", err)
	}
	_ = os.Remove(s.multipartPath(bucket))
	return object, true, nil
}

func (s *FileBucketStore) AbortMultipartUpload(ctx context.Context, bucket string, key string, uploadID string) (bool, error) {
	if err := ctx.Err(); err != nil {
		return false, err
	}
	_, ok, err := s.getMultipartUpload(ctx, bucket, key, uploadID)
	if err != nil || !ok {
		return ok, err
	}
	if err := os.RemoveAll(s.multipartUploadPath(bucket, uploadID)); err != nil {
		return true, fmt.Errorf("abort multipart upload: %w", err)
	}
	_ = os.Remove(s.multipartPath(bucket))
	return true, nil
}

func (s *FileBucketStore) ListMultipartUploads(ctx context.Context, bucket string) ([]MultipartUpload, bool, error) {
	if err := ctx.Err(); err != nil {
		return nil, false, err
	}
	if _, ok, err := s.GetBucket(ctx, bucket); err != nil {
		return nil, false, err
	} else if !ok {
		return nil, false, nil
	}
	entries, err := os.ReadDir(s.multipartPath(bucket))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, true, nil
		}
		return nil, false, fmt.Errorf("read multipart uploads: %w", err)
	}
	uploads := make([]MultipartUpload, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		var upload MultipartUpload
		if err := readJSONFile(filepath.Join(s.multipartPath(bucket), entry.Name(), "upload.json"), &upload); err != nil {
			return nil, false, err
		}
		uploads = append(uploads, upload)
	}
	sort.Slice(uploads, func(i, j int) bool {
		if uploads[i].Key == uploads[j].Key {
			return uploads[i].UploadID < uploads[j].UploadID
		}
		return uploads[i].Key < uploads[j].Key
	})
	return uploads, true, nil
}

func (s *FileBucketStore) getMultipartUpload(ctx context.Context, bucket string, key string, uploadID string) (MultipartUpload, bool, error) {
	if err := validateUploadID(uploadID); err != nil {
		return MultipartUpload{}, false, err
	}
	if err := s.requireBucketAndKey(ctx, bucket, key); err != nil {
		return MultipartUpload{}, false, err
	}
	var upload MultipartUpload
	err := readJSONFile(filepath.Join(s.multipartUploadPath(bucket, uploadID), "upload.json"), &upload)
	if err != nil {
		if os.IsNotExist(err) {
			return MultipartUpload{}, false, nil
		}
		return MultipartUpload{}, false, err
	}
	if upload.Key != key {
		return MultipartUpload{}, false, nil
	}
	return upload, true, nil
}

func (s *FileBucketStore) readMultipartParts(bucket string, uploadID string) ([]MultipartPart, error) {
	entries, err := os.ReadDir(filepath.Join(s.multipartUploadPath(bucket, uploadID), "parts"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("read multipart parts: %w", err)
	}
	parts := make([]MultipartPart, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		var part MultipartPart
		if err := readJSONFile(filepath.Join(s.multipartUploadPath(bucket, uploadID), "parts", entry.Name(), "part.json"), &part); err != nil {
			return nil, err
		}
		parts = append(parts, part)
	}
	sort.Slice(parts, func(i, j int) bool {
		return parts[i].PartNumber < parts[j].PartNumber
	})
	return parts, nil
}

File: services/s3/src/store_multipart.rs
//! The multipart-upload plane on `FileBucketStore`: create/upload-part/list/
//! complete/abort and the per-bucket upload listing. Ported 1:1 from the legacy
//! `store_multipart.rs`. The on-disk layout is
//! `multipart/<uploadId>/{upload.json, parts/<00001>/{part.json, body}}`.

use crate::hashes::{md5_etag, multipart_etag, validate_content_md5, ContentMd5Error};
use crate::model::{MultipartPart, MultipartUpload, Object};
use crate::objops::{
    clean_metadata, clean_server_side_encryption, CreateMultipartUploadInput, PutObjectInput,
};
use crate::store::{remove_dir_all_ignoring_missing, FileBucketStore, Result, StoreError};
use crate::validation::valid_upload_id;
use std::fs;
use std::io;
use std::path::Path;

fn read_json<T: serde::de::DeserializeOwned>(path: &Path) -> Result<Option<T>> {
    match fs::read(path) {
        Ok(data) => Ok(Some(
            serde_json::from_slice(&data).map_err(|e| StoreError::Io(io::Error::other(e)))?,
        )),
        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(StoreError::Io(e)),
    }
}

fn missing(what: &str) -> StoreError {
    StoreError::Io(io::Error::new(io::ErrorKind::NotFound, what.to_string()))
}

impl FileBucketStore {
    /// Begins a multipart upload, allocating an upload ID and writing
    /// `upload.json`. Defaults the content type to `application/octet-stream`.
    pub fn create_multipart_upload(
        &self,
        input: CreateMultipartUploadInput,
    ) -> Result<MultipartUpload> {
        self.require_bucket_and_key(&input.bucket, &input.key)?;
        let mut upload = MultipartUpload {
            bucket: input.bucket.clone(),
            key: input.key.clone(),
            upload_id: self.new_version_id(),
            created_at: self.now(),
            content_type: input.content_type.clone(),
            content_encoding: input.content_encoding.clone(),
            cache_control: input.cache_control.clone(),
            content_disposition: input.content_disposition.clone(),
            metadata: clean_metadata(&input.metadata),
            encryption: clean_server_side_encryption(&input.encryption),
        };
        if upload.content_type.is_empty() {
            upload.content_type = "application/octet-stream".to_string();
        }
        let path = self.multipart_upload_path(&input.bucket, &upload.upload_id);
        fs::create_dir_all(path.join("parts"))?;
        Self::write_json(&path.join("upload.json"), &upload)?;
        Ok(upload)
    }

    /// Uploads one part (1..=10000), writing its body and `part.json`.
    pub fn upload_part(
        &self,
        bucket: &str,
        key: &str,
        upload_id: &str,
        part_number: i64,
        body: &[u8],
        content_md5: &str,
    ) -> Result<MultipartPart> {
        if !(1..=10000).contains(&part_number) {
            return Err(StoreError::InvalidPartNumber);
        }
        let upload = self.get_multipart_upload(bucket, key, upload_id)?;
        match upload {
            Some(u) if u.key == key => {}
            _ => return Err(StoreError::MultipartUploadNotExist),
        }
        match validate_content_md5(content_md5, body) {
            Ok(()) => {}
            Err(ContentMd5Error::Invalid) => return Err(StoreError::InvalidContentMd5),
            Err(ContentMd5Error::Mismatch) => return Err(StoreError::ContentMd5Mismatch),
        }
        let part = MultipartPart {
            part_number,
            etag: md5_etag(body),
            size: body.len() as i64,
            last_modified: self.now(),
        };
        let path = self.multipart_part_path(bucket, upload_id, part_number);
        fs::create_dir_all(&path)?;
        fs::write(path.join("body"), body)?;
        Self::write_json(&path.join("part.json"), &part)?;
        Ok(part)
    }

    /// Lists an upload's parts (sorted by part number) with its metadata. `None`
    /// if the upload does not exist.
    pub fn list_parts(
        &self,
        bucket: &str,
        key: &str,
        upload_id: &str,
    ) -> Result<Option<(MultipartUpload, Vec<MultipartPart>)>> {
        let upload = match self.get_multipart_upload(bucket, key, upload_id)? {
            Some(u) => u,
            None => return Ok(None),
        };
        let parts = self.read_multipart_parts(bucket, upload_id)?;
        Ok(Some((upload, parts)))
    }

    /// Concatenates the listed parts into the final object (stamping the
    /// composite multipart ETag) and removes the upload. `None` if the upload is
    /// absent; `InvalidPart` if a listed part is missing.
    pub fn complete_multipart_upload(
        &self,
        bucket: &str,
        key: &str,
        upload_id: &str,
        part_numbers: &[i64],
    ) -> Result<Option<Object>> {
        let upload = match self.get_multipart_upload(bucket, key, upload_id)? {
            Some(u) => u,
            None => return Ok(None),
        };
        let mut combined: Vec<u8> = Vec::new();
        let mut part_etags: Vec<String> = Vec::with_capacity(part_numbers.len());
        for &part_number in part_numbers {
            let part_path = self.multipart_part_path(bucket, upload_id, part_number);
            let part: MultipartPart = read_json(&part_path.join("part.json"))?
                .ok_or(StoreError::InvalidPart(part_number))?;
            let body = match fs::read(part_path.join("body")) {
                Ok(b) => b,
                Err(e) if e.kind() == io::ErrorKind::NotFound => {
                    return Err(StoreError::InvalidPart(part_number));
                }
                Err(e) => return Err(StoreError::Io(e)),
            };
            combined.extend_from_slice(&body);
            part_etags.push(part.etag);
        }
        let mut object = self.put_object(PutObjectInput {
            bucket: upload.bucket.clone(),
            key: upload.key.clone(),
            body: combined.clone(),
            content_type: upload.content_type.clone(),
            content_encoding: upload.content_encoding.clone(),
            cache_control: upload.cache_control.clone(),
            content_disposition: upload.content_disposition.clone(),
            metadata: upload.metadata.clone(),
            encryption: upload.encryption.clone(),
            ..Default::default()
        })?;
        object.etag = multipart_etag(&part_etags);
        let object_path = self.object_path(&upload.bucket, &upload.key);
        Self::write_json(&object_path.join("object.json"), &object)?;
        if !object.version_id.is_empty() {
            self.write_object_version(&object_path, &object, &combined)?;
        }
        remove_dir_all_ignoring_missing(&self.multipart_upload_path(bucket, upload_id))?;
        let _ = fs::remove_dir(self.multipart_path(bucket));
        Ok(Some(object))
    }

    /// Aborts an upload, discarding its parts. `true` if it existed.
    pub fn abort_multipart_upload(&self, bucket: &str, key: &str, upload_id: &str) -> Result<bool> {
        if self.get_multipart_upload(bucket, key, upload_id)?.is_none() {
            return Ok(false);
        }
        remove_dir_all_ignoring_missing(&self.multipart_upload_path(bucket, upload_id))?;
        let _ = fs::remove_dir(self.multipart_path(bucket));
        Ok(true)
    }

    /// Lists a bucket's in-progress uploads, sorted by key then upload ID. `None`
    /// if the bucket does not exist.
    pub fn list_multipart_uploads(&self, bucket: &str) -> Result<Option<Vec<MultipartUpload>>> {
        if self.get_bucket(bucket)?.is_none() {
            return Ok(None);
        }
        let entries = match fs::read_dir(self.multipart_path(bucket)) {
            Ok(e) => e,
            Err(e) if e.kind() == io::ErrorKind::NotFound => return Ok(Some(Vec::new())),
            Err(e) => return Err(StoreError::Io(e)),
        };
        let mut uploads = Vec::new();
        for entry in entries {
            let entry = entry?;
            if !entry.file_type()?.is_dir() {
                continue;
            }
            let upload: MultipartUpload = read_json(&entry.path().join("upload.json"))?
                .ok_or_else(|| missing("read multipart upload metadata"))?;
            uploads.push(upload);
        }
        uploads.sort_by(|a, b| match a.key.cmp(&b.key) {
            std::cmp::Ordering::Equal => a.upload_id.cmp(&b.upload_id),
            other => other,
        });
        Ok(Some(uploads))
    }

    fn get_multipart_upload(
        &self,
        bucket: &str,
        key: &str,
        upload_id: &str,
    ) -> Result<Option<MultipartUpload>> {
        if !valid_upload_id(upload_id) {
            return Err(StoreError::InvalidUploadId);
        }
        self.require_bucket_and_key(bucket, key)?;
        let upload: MultipartUpload = match read_json(
            &self
                .multipart_upload_path(bucket, upload_id)
                .join("upload.json"),
        )? {
            Some(u) => u,
            None => return Ok(None),
        };
        if upload.key != key {
            return Ok(None);
        }
        Ok(Some(upload))
    }

    fn read_multipart_parts(&self, bucket: &str, upload_id: &str) -> Result<Vec<MultipartPart>> {
        let parts_dir = self.multipart_upload_path(bucket, upload_id).join("parts");
        let entries = match fs::read_dir(&parts_dir) {
            Ok(e) => e,
            Err(e) if e.kind() == io::ErrorKind::NotFound => return Ok(Vec::new()),
            Err(e) => return Err(StoreError::Io(e)),
        };
        let mut parts = Vec::new();
        for entry in entries {
            let entry = entry?;
            if !entry.file_type()?.is_dir() {
                continue;
            }
            let part: MultipartPart = read_json(&entry.path().join("part.json"))?
                .ok_or_else(|| missing("read multipart part metadata"))?;
            parts.push(part);
        }
        parts.sort_by(|a, b| a.part_number.cmp(&b.part_number));
        Ok(parts)
    }
}
